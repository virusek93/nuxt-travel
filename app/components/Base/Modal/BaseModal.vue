<script lang="ts" setup>
const dialogRef = ref<HTMLDialogElement>()
const props = defineProps<{ isModalOpen: boolean }>()
defineEmits(['close'])
watchEffect(() => {
  if (!dialogRef.value) {
    return
  }
  if (props.isModalOpen) {
    dialogRef.value.showModal()
    return
  }
  dialogRef.value.close()
})
</script>

<template>
  <dialog
    ref="dialogRef"
    class="rounded bg-white p-5 backdrop:bg-black/50"
  >
    <div
      class="mb-4 flex items-center"
      data-test="content"
    >
      <slot name="title" />
      <button
        type="button"
        class="ml-auto flex min-h-4 min-w-4 items-center transition-opacity hover:opacity-50"
        @click="$emit('close')"
      >
        X
      </button>
    </div>
    <slot />
  </dialog>
</template>
