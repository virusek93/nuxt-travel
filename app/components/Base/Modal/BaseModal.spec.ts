import { describe, expect, it, beforeAll, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import BaseModal from '~/components/Base/Modal/BaseModal.vue'

const props: InstanceType<typeof BaseModal>['$props'] = {
  isModalOpen: false,
}

describe('BaseModal', () => {
  beforeAll(() => {
    HTMLDialogElement.prototype.show = vi.fn()
    HTMLDialogElement.prototype.showModal = vi.fn()
    HTMLDialogElement.prototype.close = vi.fn()
  })
  it('Should render and match the snapshot', () => {
    const wrapper = mount(BaseModal, { props })
    expect(wrapper.element).toMatchSnapshot()
  })
  it('Should call showModal on dialog element when isModalOpen sets to true', async () => {
    const wrapper = mount(BaseModal, { props })
    // @ts-expect-error vue-closed composition API
    const spy = vi.spyOn(wrapper.vm.dialogRef, 'showModal')
    await wrapper.setProps({ isModalOpen: true })
    expect(spy).toHaveBeenCalled()
  })
  it('Should call close on dialog element when isModalOpen sets to false', async () => {
    const wrapper = mount(BaseModal, { props: { isModalOpen: true } })
    // @ts-expect-error vue-closed composition API
    const spy = vi.spyOn(wrapper.vm.dialogRef, 'close')
    await wrapper.setProps({ isModalOpen: false })
    expect(spy).toHaveBeenCalled()
  })
  it('Should render slot details', () => {
    const defaultSlotData = 'Test default'
    const wrapper = mount(BaseModal, {
      props: { isModalOpen: true },
      slots: { default: defaultSlotData },
    })
    expect(wrapper.text().includes(defaultSlotData))
  })
})
