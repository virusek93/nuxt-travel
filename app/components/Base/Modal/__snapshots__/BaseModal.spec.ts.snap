// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseModal > Should render and match the snapshot 1`] = `
<dialog
  class="rounded bg-white p-5 backdrop:bg-black/50"
>
  <div
    class="mb-4 flex items-center"
    data-test="content"
  >
    
    
    <button
      class="ml-auto flex min-h-4 min-w-4 items-center transition-opacity hover:opacity-50"
      type="button"
    >
       X 
    </button>
  </div>
  
  
</dialog>
`;
